src/Datasource/EntityTrait.php

```diff
@@ -437,12 +437,16 @@ public function toArray()
         $result = [];
         foreach ($this->visibleProperties() as $property) {
             $value = $this->get($property);
-            if (is_array($value) && isset($value[0]) && $value[0] instanceof EntityInterface) {
+            if (is_array($value)) {
                 $result[$property] = [];
                 foreach ($value as $k => $entity) {
-                    $result[$property][$k] = $entity->toArray();
+                    if (method_exists($entity, 'toArray')) {
+                        $result[$property][$k] = $entity->toArray();
+                    } else {
+                        $result[$property][$k] = $entity;
+                    }
                 }
-            } elseif ($value instanceof EntityInterface) {
+            } elseif (method_exists($value, 'toArray')) {
```

Comment:
`instanceof` is only slightly faster than `method_exists`, to the point where I don't think it matters unless you're doing this thousands of times in a request - in which case *maybe* you should be doing something else in that request?